libnymea-core/hardware/zwave/zwavehardwareresourceimplementation.cpp

```diff
@@ -0,0 +1,158 @@
+#include "zwavehardwareresourceimplementation.h"
+
+#include "zwave/zwavemanager.h"
+#include "zwave/zwavenetwork.h"
+
+#include "loggingcategories.h"
+Q_DECLARE_LOGGING_CATEGORY(dcZWave)
+
+namespace nymeaserver
+{
+
+
+ZWaveHardwareResourceImplementation::ZWaveHardwareResourceImplementation(ZWaveManager *zwaveManager, QObject *parent):
+    ZWaveHardwareResource(parent),
+    m_zwaveManager(zwaveManager)
+{
+    connect(m_zwaveManager, &ZWaveManager::networkStateChanged, this, &ZWaveHardwareResourceImplementation::onNetworkStateChanged);
+    connect(m_zwaveManager, &ZWaveManager::nodeInitialized, this, &ZWaveHardwareResourceImplementation::onNodeInitialized);
+    connect(m_zwaveManager, &ZWaveManager::nodeRemoved, this, &ZWaveHardwareResourceImplementation::onNodeRemoved);
+//    connect(m_zigbeeManager, &ZigbeeManager::availableChanged, this, &ZigbeeHardwareResourceImplementation::onZigbeeAvailableChanged);
+
+}
+
+bool ZWaveHardwareResourceImplementation::available() const
+{
+    return m_zwaveManager->available();
+}
+
+bool ZWaveHardwareResourceImplementation::enabled() const
+{
+    return m_zwaveManager->enabled();
+}
+
+void ZWaveHardwareResourceImplementation::setEnabled(bool enabled)
+{
+    m_zwaveManager->setEnabled(enabled);
+}
+
+void ZWaveHardwareResourceImplementation::registerHandler(ZWaveHandler *handler, HandlerType type)
+{
+    qCDebug(dcZWave()) << "Registering new Z-Wave handler" << handler->name() << "with type" << type;
+    m_handlers.insert(type, handler);
+}
+
+ZWaveNode *ZWaveHardwareResourceImplementation::claimNode(ZWaveHandler *handler, const QUuid &networkUuid, quint8 nodeId)
+{
+    if (!m_handlers.values().contains(handler)) {
+        qCWarning(dcZWave()) << "Handler" << handler->name() << "is not registered. Not allowing node to be claimed.";
+        return nullptr;
+    }
+
+    ZWaveNetwork *network = m_zwaveManager->network(networkUuid);
+    if (!network) {
+        qCWarning(dcZWave()) << "Network" << networkUuid << "not found.";
+        return nullptr;
+    }
+
+    ZWaveNode *node = network->node(nodeId);
+    if (!node) {
+        qCWarning(dcZWave()) << "Node with ID" << nodeId << "not found in ZWave network" << networkUuid.toString();
+        return nullptr;
+    }
+
+    if (m_nodeHandlers.contains(node) && m_nodeHandlers.value(node) != handler) {
+        qCWarning(dcZWave()) << "Node with ID" << nodeId << "is already claimed by another handler (" << m_nodeHandlers.value(node)->name() << "). Not allowing node to be reclaimed.";
+        return nullptr;
+    }
+
+    m_nodeHandlers[node] = handler;
+    return node;
+}
+
+void ZWaveHardwareResourceImplementation::removeNodeFromNetwork(const QUuid &networkUuid, ZWaveNode *node)
+{
+    qCritical() << "remove node" << networkUuid << node;
```

Comment:
right... no, this won't work with Z-Wave... Removing a node needs to use the same procedure as adding a node. Dropped the method.

This will actually still allow the user to remove a thing, but it will come back after the next restart... Not much I can do about that tho currently as we'd need to introduce the canDelete propety in the thingClass